Runner: periodically check if there's a new version available on Github and offer a visit

yuyoyuppe opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
- add a worker thread which periodically (currently once per 24h) checks if there's a new release available on GitHub
- it checks if the current version _string_, obtained with `get_product_version()`, isn't equal to the `tag_name` (the actual git tag) from the latest release obtained using GitHub API, so we must be **VERY CAREFUL** when submitting stuff on [/releases](https://github.com/microsoft/PowerToys/releases) pages, because users will always be getting a notification if the tag doesn't match
- if a new version is found, we offer a user to go to its download page, [e.g. for v0.14.1](https://github.com/microsoft/PowerToys/releases/tag/v0.14.1)
- the last time we've checked for updates or showed a user the update dialog is tracked in a separate file at `%localappdata%\Microsoft\PowerToys\update_state.json` to not mix it with general settings and be able to save/load it safely anytime
- `VERSION_BUILD` isn't used in `get_product_version` anymore, also it now returns a version prepended with `'v'` to much with git tags notation we have so far (e.g. v0.14.1). We still have another zero in `PRODUCTVERSION`/`FILEVERSION` macros to [make WinAPI happy](https://docs.microsoft.com/en-us/windows/win32/menurc/versioninfo-resource)

Once we've added support for toast notifications in non-packaged applications, we can use a toast instead.

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [x] Applies to #696

## Validation Steps Performed
- launched (non-)packaged PTs and played w/ values from update_state.json file

## Screenshot
![image](https://user-images.githubusercontent.com/1828123/75047672-39f13680-54d8-11ea-8bd9-bdf87a659aeb.png)

yuyoyuppe committed:
common: always use 0 for build number and make get_product_version return "vX.X.X"

yuyoyuppe committed:
runner: periodically check if there's a new version available on github and offer a visit

enricogior commented:
We may consider do some parsing when the current version doesn't match the github latest release, to do a more reliable comparison.
We need to verify if creating a draft release triggers the notification.

ivan100sic commented:
I really wouldn't want a message box popping up in the middle of a game

crutkas commented:
Have we verified that would happen with the type of msg box we are using?

PrzemyslawTusinski committed:
Now updater is comparing versions instead of comparing thier strings

enricogior commented:
@crutkas @ivan100sic 
we are replacing the message box with a toast notification. The work to support toast notification in "MSI mode" is in progress.

yuyoyuppe committed:
notifications: add support for unpackaged apps and protocol activation

yuyoyuppe commented:
### Description of the work done in the recently pushed commits

#### Notifications library improvements
- added support for toasts in MSI version of PT (thanks @stefansjfw for initial work)
- added support for links in toasts
- refactored COM activation libraries
- XML input is now escaped
- improved XML handling

Also, background-activated toasts for MSI version use the same dispatcher (see handler_functions.cpp) as the MSIX to allow code-sharing. For that, we must use PowerToys.exe as an initial local COM server and load the Notifications.dll separately. However, doing this work I've found that it's actually [impossible to launch a URI from a background task](https://stackoverflow.com/questions/48317164/launch-an-uwp-app-from-background-task), so I've added `protocol` activation type support.


#### Screenshot 
![image](https://user-images.githubusercontent.com/1828123/75284693-55d14100-5826-11ea-9791-2bc81b878ed7.png)

enricogior commented:
@PrzemyslawTusinski we should add unit tests for VersionHelper.cpp.

enricogior commented:
MSI:

![image](https://user-images.githubusercontent.com/3206696/75333914-40403380-5887-11ea-923f-bdac63654c61.png)

MSIX (Experimental):

![image](https://user-images.githubusercontent.com/3206696/75333682-db84d900-5886-11ea-9297-4e2f1c428e61.png)

enricogior commented:
We have only a minor issue when clicking on the "Visit" button: if the default browser is already open and not in the foreground, it will remain not in the foreground.
I guess we can live with that for this release, and investigate if we can workaround it in 0.16.

enricogior commented on src/modules/powerrename/UWPui/PowerRenameUWPUI.cpp:
```diff
@@ -4,6 +4,7 @@
 #include <CLSID.h>
 #include <PowerRenameExt.h>
 #include <common.h>
+#include <com_object_factory.h>
```
Let's revert this change to PowerRenameUWPUI.cpp and open an issue to track it.

yuyoyuppe replied:
https://github.com/microsoft/PowerToys/issues/1367

enricogior commented on src/runner/main.cpp:
```diff
@@ -165,8 +214,44 @@ int runner(bool isProcessElevated)
     return result;
 }
 
+enum class SpecialMode
```
Can you add a comment to explain what SpecialMode is?

enricogior commented on src/common/UnitTests-CommonLib/msi_to_msix_upgrade_lib.Tests.cpp:
```diff
@@ -0,0 +1,93 @@
+
+using namespace Microsoft::VisualStudio::CppUnitTestFramework;
+
+namespace UnitTestsmsitomsixupgrade
```
Rename `UnitTestsmsitomsixupgrade` to `UnitTestsVersionHelper`

enricogior commented on src/common/UnitTests-CommonLib/msi_to_msix_upgrade_lib.Tests.cpp:
```diff
@@ -0,0 +1,93 @@
+    const int MINOR_VERSION_12 = 12;
+    const int REVISION_VERSION_0 = 0;
+
+    TEST_CLASS(UnitTestsmsitomsixupgrade)
```
Rename `UnitTestsmsitomsixupgrade` to `UnitTestsVersionHelper`

yuyoyuppe commented:
Rebasing changes to get a sensible commit history.

crutkas commented:
How about ... "An update to PowerToys is available.  Visit our GitHub page to get vXXXXX.".  The current one is very verbose.

yuyoyuppe commented:
@crutkas done 🙂
![image](https://user-images.githubusercontent.com/1828123/75369541-b7e87f80-58d4-11ea-9b91-941140687877.png)
